Full OnGasConsumed loop and added `GasChangeReason`

maoueh opened the pull request:
With this change, the full gas loop of a transaction can be tracked going from initial balance (`gasLimit`) back down to 0.

I validated on my regression test suite that those conditions applied to all transactions I have:
- First `OnGasConsumed` is always going from 0 -> `trx.gasLimit`
- Last `OnGasConsumed` "new value" is always 0
- trx.gasLimit - last balance of last `OnGasConsumed` is equal to `trx.gasUsed`

The addition of `reason` make it possible for a logger to filter out unwanted signal, for example one could want to drop all OpCode related gas change and only cares about "transaction" level. The reason is also a good thing for visibility and for creating powerful debug tools for Ethereum transaction execution.

Also, I would like that we change the `OnGasConsumed` logic from `OnGasConsumed(actual, cost uint64, reason)` to `OnGasConsumed(old, new uint64, reason)`. With the new tracing I've done, we now have negative `cost` that needs to be passed around but the `cost` is `uint64` so it creates overflow. Everything works fine if you record the new/old value `gas-cost` as the final value goes back to a valid range. But it's weird for consumer that would log `cost` that it's `18127127187219...`. An alternative would be to accept a `int64` for the cost which is also valid, not full coverage theorically but I don't see why an operation would cost more than i64 max value. Maybe it make more sense than to have signature `OnGasConsumed(gas uint64, cost int64, reason)`, I'm fine with it also, I leave the decision to you.

maoueh committed:
Full OnGasConsumed loop and added `GasChangeReason`

With this change, the full gas loop of a transaction can be tracked going from initial balance (`gasLimit`) back down to 0.

I validated on my regression test suite that those conditions applied to all transactions I have:
- First `OnGasConsumed` is always going from 0 -> `trx.gasLimit`
- Last `OnGasConsumed` "new value" is always 0
- trx.gasLimit - last balance of last `OnGasConsumed` is equal to `trx.gasUsed`

The addition of `reason` make it possible for a logger to filter out unwanted signal, for example one could want to drop all OpCode related gas change and only cares about "transaction" level. The reason is also a good thing for visibility and for creating powerful debug tools for Ethereum transaction execution.

Also, I would like that we change the `OnGasConsumed` logic from `OnGasConsumed(actual, cost uint64, reason)` to `OnGasConsumed(old, new uint64, reason)`. With the new tracing I've done, we now have negative `cost` that needs to be passed around but the `cost` is `uint64` so it creates overflow. Everything works fine if you record the new/old value `gas-cost` as the final value goes back to a valid range. But it's weird for consumer that would log `cost` that it's `18127127187219...`. An alternative would be to accept a `int64` for the cost which is also valid, not full coverage theorically but I don't see why an operation would cost more than i64 max value. Maybe it make more sense than to have signature `OnGasConsumed(gas uint64, cost int64, reason)`, I'm fine with it also, I leave the decision to you.

s1na commented:
Hm this is not looking good to do `-cost`. What do you think about having 2 methods: `GasCharge` and `GasBump` (e.g.) instead? Otherwise I agree about doing `OnGasChange(new, value, reason)` and let consumers figure out if it was an increase or decrease if they care.

maoueh commented:
I'm fine with both approach.

The single method seems more appealing IMO however from a tracer perspective, less cognitive thinking than with having 2 methods, but it has the downsize of not showing "the cost" directly.

If we go with 2 methods, I don't personally like `GasCharge/Bump`, it think it's the `GasBump` part that triggers that to me, the `GasCharge` feels good.

So maybe: `GasCharge/GasRefund` (`refund` could be misleading), `GasDebit/GasCredit`, `GasIncrease/GasDecrease`, `GasConsumed/GasReturned` (`returned` could be misleading). 

I leave you the final decision.

maoueh committed:
Change `OnGasConsumed(gas, cost uint64, reason)` to `OnGasChange(old, new uint64, reason)`

This way, we avoid having a `cost` that is negative which does not make sense for a `uint64`. Having the `old, new` also yields correct value and the delta can then be negative and be holded in a `int64`.

maoueh commented:
I've updated this PR to use signature `OnGasChange(old, new uint64, reason)` so that `-cost` does not happen anymore. Let me know if you would prefer something else (either `OnGasChange(gas, cost int64, reason)` or `OnGasIncrease/OnGasDecrease`)

s1na commented on core/vm/logger.go:
```diff
@@ -43,5 +43,52 @@ type EVMLogger interface {
+	// always exactly one of those per transaction
+	GasChangeIntrinsicGas
+
+	// PR Review: `GasChangeContractCreation/2` are actually the EIP150 burn cost of CREATE/CREATE2 respectively.
```
Hm no that is the gas provided to create for its execution, i.e. for contract initialization. I think `GasChangeCreate/2` is good.

maoueh replied:
Perfect

s1na commented on core/vm/logger.go:
```diff
@@ -43,5 +43,52 @@ type EVMLogger interface {
+	GasInitialBalance
+	// GasRefunded is the amount of gas that will be refunded to the caller for data returned to the chain
+	// PR Review: Is that the right description? Called in core/state_transition.go#StateTransition.refundGas
+	GasRefunded
```
`GasRefunded` is the sum of all refunds which happened during the tx execution (e.g. storage slot being cleared).

I see the total that will be returned to the caller is tracked via `GasBuyBack`. Is this distinction important? I would think if we want to track the refund values tracking each of them is more valuable than the sum at the end? i.e. emit an event at the site when storage is cleared etc.

maoueh replied:
> I see the total that will be returned to the caller is tracked via GasBuyBack. Is this distinction important?

Good question actually. I do feel its important but it's a personal opinion I'm sharing here as we didn't track those before.

My thinking here is that gas refund has been used speculatively for "gas value accrual" in the past so emitting this event make sense to me, it will be easy to know how much refund has been given back to you.

If it was "blended" with `GasBuyBack`, I'm unsure it would be possible to extract the refund value. I might be wrong and maybe. it could be computed/inferred based on surrounding gas changes, I do still think an explicit version is better.

I do get that it could be blended in indeed.

> I would think if we want to track the refund values tracking each of them is more valuable than the sum at the end

I don't think so, here is why. Correct me if I'm wrong, when the storage is cleared and the refund accumulate, the gas counter is not incremented meaning the gas value the execution sees don't contain the refund.

It's would be falsely IMO to emit a `GasChange` at those points because the execution don't account for them yet. I think the logger should faithfully trace the execution as it happens.

Let me know what do you think.

s1na replied:
> the gas counter is not incremented meaning the gas value the execution sees don't contain the refund.

Ah yes that's a good point.

> I think the logger should faithfully trace the execution as it happens.

Agreed.

>  it will be easy to know how much refund has been given back to you.

Ok then we keep a separate reason for refunds 👍

s1na commented on core/vm/logger.go:
```diff
@@ -43,5 +43,52 @@ type EVMLogger interface {
+	GasChangeStorageColdAccess
+
+	// GasChangeCallLeftOverRefunded is the amount of gas that will be refunded to the caller after the execution of the call, if there is left over at the end of execution
+	GasChangeCallLeftOverRefunded
```
Can you please clarify in the comments the distinction between LeftOverRefunded and BuyBack?

maoueh replied:
Here the new descriptions:

```
	// GasChangeTxBuyBack is the amount of gas that will be bought back by the chain and returned in Wei to the caller at the very
	// end of the transaction's execution. There is only one such gas change per transaction.
	GasChangeTxBuyBack
```

```
	// GasChangeCallLeftOverRefunded is the amount of gas that will be refunded to the caller after the execution of the call, if
	// there is left over at the end of call's execution. This can change can happen multiple times within a single transaction as
	// each call is independent of each other.
	GasChangeCallLeftOverRefunded
```

I've also normalized all the names so they read either `GasChangeTx` or `GasChangeCall` easily picking if it's a tx level gas change or a call level gas change.

I've also added the checks to not emit `GasChange` if on refunds there is 0 gas to return or on buy back there is no gas to buy back.

s1na replied:
I'm confused about your terminology here. As I can see `GasChangeTxBuyBack` is definitely called more than once per tx, it's indeed called once per call. Same with `GasChangeTxInitialBalance`. They are called at the same time as `CaptureEnter` and `CaptureExit` in core/vm/evm.go.

maoueh replied:
You are right, I did this in a haste and I completely miss the nuance of the implementation indeed. My idea is to split has change related to `state_transition` from those done at the `call` level. 

My latest PR fixes this introducing correct split for `Tx/Call`. You can check again. I've also again improved the documentation of each gas change.

maoueh replied:
This looks like this currently:

```
# Pure Transfer
Tx
    - GasChangeTxInitialBalance +21000 (21000)
    - GasChangeTxIntrinsicGas -21000 (0)
    + Call 0 (root)
      End
```

```
# Transfer via a sub-call
Tx
    - GasChangeTxInitialBalance +75000 (75000)
    - GasChangeTxIntrinsicGas Cost -21432 (53568)
    + Call 0 (root) 
        - GasChangeCallInitialBalance +53568 (53568)
        - GasChangeCallOpCode [...] ...
        - GasChangeCallOpCode [CALL] -9700 (43491)
        + Call 1 (child)
            - GasChangeCallInitialBalance +2300 (2300)
            - GasChangeCallLeftOverReturned -2300 (0)
          End
        - GasChangeCallLeftOverRefunded +2300 (45791)
        - GasChangeCallLeftOverReturned -45744 (0)
      End
    - GasChangeTxBuyBack -45744 (0)
```

maoueh committed:
Normalized `GasChange` constats, emit `Refund/BuyBack` only if doing something, updated comments

maoueh committed:
Merge remote-tracking branch 's1na/extended-tracer' into feature/gas-full-cycle-and-reason

# Conflicts:
#	eth/tracers/logger/access_list_tracer.go
#	eth/tracers/logger/logger.go
#	eth/tracers/logger/logger_json.go

maoueh committed:
Merge remote-tracking branch 's1na/extended-tracer' into feature/gas-full-cycle-and-reason

maoueh committed:
Fixed wrong split of trx and call

maoueh committed:
Inverted `GasChangeCallLeftOverRefunded` and `GasChangeCallLeftOverReturned`

maoueh committed:
Fixed some missing gas change to wrap around full call

maoueh committed:
Do not record `GasChangeCallLeftOverReturned` gas change if there is no gas to return

maoueh committed:
This refactors now beefy `*EVM` call tracing to be much less duplicated

maoueh commented:
Also, the EVM tracing is starting to be really big, how about refactoring it to use a common code on `*EVM` directly. Here how it looks like: https://github.com/streamingfast/go-ethereum/pull/10.

Let me know what you think.

s1na commented on core/vm/instructions.go:
```diff
@@ -612,6 +612,11 @@ func opCreate(pc *uint64, interpreter *EVMInterpreter, scope *ScopeContext) ([]b
 	scope.Stack.push(&stackvalue)
+
+	if interpreter.evm.Config.Tracer != nil {
+		interpreter.evm.Config.Tracer.OnGasChange(scope.Contract.Gas, scope.Contract.Gas+returnGas, GasChangeCallLeftOverRefunded)
```
Ah I see so from within the call we "return leftover gas" and in the parent call we "refund the leftover", makes sense.

s1na commented on core/vm/logger.go:
```diff
@@ -43,5 +43,57 @@ type EVMLogger interface {
+	// GasChangeCallLeftOverRefunded is the amount of gas that will be refunded to the call after the child call execution it
+	// executed completed. This value is always positive as we are giving gas back to the you, the left over gas of the child.
+	GasChangeCallLeftOverRefunded
+	// GasChangeCallContractCreation is the amount of gas that will be burned for a CREATE, today controlled by EIP150 rules.
```
Please remove the EIP150 part. That only regulates one part of the gas fee (63/64th rule). The create gas calculation has been changed several times since in other ways.

maoueh committed:
More review of gas changes

maoueh committed:
Merge branch 'feature/refactored-evm-call-tracing-code' into feature/gas-full-cycle-and-reason

s1na reviewed (approved):
LGTM
